pkg/manager/member/tidb_discovery_manager.go

```diff
@@ -45,29 +48,49 @@ func NewTidbDiscoveryManager(deps *controller.Dependencies) TidbDiscoveryManager
 	return &realTidbDiscoveryManager{deps: deps}
 }
 
-func (m *realTidbDiscoveryManager) Reconcile(tc *v1alpha1.TidbCluster) error {
+func (m *realTidbDiscoveryManager) Reconcile(obj runtime.Object) error {
+	metaObj, ok := obj.(metav1.Object)
+	if !ok {
+		return fmt.Errorf("%T is not a metav1.Object", obj)
+	}
 
-	// If PD is not specified return
-	if tc.Spec.PD == nil {
+	meta, _ := getDiscoveryMeta(metaObj, controller.DiscoveryMemberName)
```

Comment:
What about moving this after L87 so that in this function we can assume that no unknown type case for the metaObj?

Reply:
updated in https://github.com/pingcap/tidb-operator/pull/3914/commits/92b8de1bd874ffe89a5f4496879939ca275bb5ef